Error in query - CALL apoc.schema.nodes()

Neo4j Vsersion : 3.5.19
Apoc Versions: 3.5.0.4 and 3.5.0.12

An issue in APOC procedure CALL apoc.schema.nodes(), It runs fine when there is NO index but It got fails with below error when there is any type of index in the database

```
Neo.ClientError.Procedure.ProcedureCallFailed
Neo.ClientError.Procedure.ProcedureCallFailed: Failed to invoke procedure `apoc.schema.nodes`: Caused by: java.lang.IncompatibleClassChangeError: Found interface org.neo4j.storageengine.api.schema.PopulationProgress, but class was expected
```
![IndexOutput](https://user-images.githubusercontent.com/31652619/86510874-c3e8ce80-bdeb-11ea-84b1-ee167a27b4d0.png)


Requesting a fix for this bug and need a updated apoc version to share with the customer please

--- a/build.gradle
+++ b/build.gradle
@@ -40,7 +40,7 @@ description = """neo4j-apoc-procedures"""
 
 ext {
     // NB: due to version.json generation by parsing this file, the next line must not have any if/then/else logic
-    neo4jVersion = "3.5.17"
+    neo4jVersion = "3.5.19"
     // instead we apply the override logic here
     neo4jVersionEffective = project.hasProperty("neo4jVersionOverride") ? project.getProperty("neo4jVersionOverride") : neo4jVersion
     testContainersVersion = '1.11.0'

--- a/src/main/java/apoc/custom/CypherProcedures.java
+++ b/src/main/java/apoc/custom/CypherProcedures.java
@@ -175,8 +175,8 @@ public class CypherProcedures {
 
         public boolean registerProcedure(String name, String statement, String mode, List<List<String>> outputs, List<List<String>> inputs, String description) {
             boolean admin = false; // TODO
-            ProcedureSignature signature = new ProcedureSignature(qualifiedName(name), inputSignatures(inputs), outputSignatures(outputs),
-                    mode(mode), admin, null, new String[0], description, null, false, true
+            ProcedureSignature signature = Signatures.createProcedureSignature(qualifiedName(name), inputSignatures(inputs), outputSignatures(outputs),
+                    mode(mode), admin, null, new String[0], description, null, false, true, false
             );
             return registerProcedure(signature, statement);
 
@@ -214,8 +214,8 @@ public class CypherProcedures {
 
         public boolean removeProcedure(String name, List<List<String>> inputs, List<List<String>> outputs, String description, String mode) {
             boolean admin = false; // TODO
-            ProcedureSignature signature = new ProcedureSignature(qualifiedName(name), inputSignatures(inputs), outputSignatures(outputs),
-                    mode(mode), admin, null, new String[0], description, null, false, true
+            ProcedureSignature signature = Signatures.createProcedureSignature(qualifiedName(name), inputSignatures(inputs), outputSignatures(outputs),
+                    mode(mode), admin, null, new String[0], description, null, false, true, false
             );
             return removeProcedure(signature);
         }

--- a/src/main/java/apoc/custom/Signatures.java
+++ b/src/main/java/apoc/custom/Signatures.java
@@ -1,17 +1,45 @@
 package apoc.custom;
 
 import apoc.util.JsonUtil;
-import org.antlr.v4.runtime.*;
-import org.neo4j.internal.kernel.api.procs.*;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CodePointCharStream;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.neo4j.internal.kernel.api.procs.DefaultParameterValue;
+import org.neo4j.internal.kernel.api.procs.FieldSignature;
+import org.neo4j.internal.kernel.api.procs.Neo4jTypes;
+import org.neo4j.internal.kernel.api.procs.ProcedureSignature;
+import org.neo4j.internal.kernel.api.procs.QualifiedName;
+import org.neo4j.internal.kernel.api.procs.UserFunctionSignature;
 import org.neo4j.procedure.Mode;
 
+import java.lang.reflect.Constructor;
 import java.util.ArrayList;
 import java.util.Collections;
 import java.util.List;
 import java.util.Map;
 import java.util.stream.Collectors;
 
-import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.*;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTAny;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTBoolean;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTDate;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTDateTime;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTDuration;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTFloat;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTGeometry;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTInteger;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTLocalDateTime;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTLocalTime;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTMap;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTNode;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTNumber;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTPath;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTPoint;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTRelationship;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTString;
+import static org.neo4j.internal.kernel.api.procs.Neo4jTypes.NTTime;
 
 public class Signatures {
 
@@ -72,7 +100,7 @@ public class Signatures {
         String warning = null; // "todo warning";
         boolean eager = false;
         boolean caseInsensitive = true;
-        return new ProcedureSignature(name, inputSignatures, outputSignature, mode, admin, deprecated, allowed, description, warning, eager, caseInsensitive);
+        return createProcedureSignature(name, inputSignatures, outputSignature, mode, admin, deprecated, allowed, description, warning, eager, caseInsensitive, false);
     }
 
     public List<String> namespace(SignatureParser.NamespaceContext namespaceContext) {
@@ -212,5 +240,60 @@ public class Signatures {
     public ProcedureSignature asProcedureSignature(String signature, String description, Mode mode) {
         SignatureParser.ProcedureContext ctx = parseProcedure(signature);
         return toProcedureSignature(ctx, description, mode);
+    }
+
+    public static ProcedureSignature createProcedureSignature(QualifiedName name,
+                                                              List<FieldSignature> inputSignature,
+                                                              List<FieldSignature> outputSignature,
+                                                              Mode mode,
+                                                              boolean admin,
+                                                              String deprecated,
+                                                              String[] allowed,
+                                                              String description,
+                                                              String warning,
+                                                              boolean eager,
+                                                              boolean caseInsensitive,
+                                                              boolean systemProcedure) {
+        try {
+            // in Neo4j 3.5.19 org.neo4j.internal.kernel.api.procs.ProcedureSignature
+            // changed the signature adding a boolean at the end and without leaving the old signature
+            // in order to maintain the backwards compatibility with version prior to 4.0.5 we use the
+            // reflection to create a new instance of the class
+            final Class<?> clazz = Class.forName("org.neo4j.internal.kernel.api.procs.ProcedureSignature");
+            final Constructor<?>[] constructors = clazz.getConstructors();
+            for (int i = 0; i < constructors.length; i++) {
+                final Constructor<?> constructor = constructors[i];
+                switch (constructor.getParameterCount()) {
+                    case 12:
+                        return (ProcedureSignature) constructor.newInstance(name,
+                                inputSignature,
+                                outputSignature,
+                                mode,
+                                admin,
+                                deprecated,
+                                allowed,
+                                description,
+                                warning,
+                                eager,
+                                caseInsensitive,
+                                systemProcedure);
+                    case 11:
+                        return (ProcedureSignature) constructor.newInstance(name,
+                                inputSignature,
+                                outputSignature,
+                                mode,
+                                admin,
+                                deprecated,
+                                allowed,
+                                description,
+                                warning,
+                                eager,
+                                caseInsensitive);
+                }
+            }
+            throw new RuntimeException("Constructor of org.neo4j.internal.kernel.api.procs.ProcedureSignature not found");
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
     }
 }

--- a/src/main/java/apoc/schema/Schemas.java
+++ b/src/main/java/apoc/schema/Schemas.java
@@ -2,26 +2,43 @@ package apoc.schema;
 
 import apoc.result.AssertSchemaResult;
 import apoc.result.ConstraintRelationshipInfo;
-import apoc.result.GraphResult;
 import apoc.result.IndexConstraintNodeInfo;
 import apoc.util.Util;
 import org.apache.commons.lang3.StringUtils;
-import org.neo4j.graphdb.*;
+import org.neo4j.graphdb.GraphDatabaseService;
+import org.neo4j.graphdb.Label;
+import org.neo4j.graphdb.RelationshipType;
 import org.neo4j.graphdb.schema.ConstraintDefinition;
 import org.neo4j.graphdb.schema.ConstraintType;
 import org.neo4j.graphdb.schema.IndexDefinition;
 import org.neo4j.graphdb.schema.Schema;
 import org.neo4j.helpers.collection.Iterables;
-import org.neo4j.internal.kernel.api.*;
+import org.neo4j.internal.kernel.api.IndexReference;
+import org.neo4j.internal.kernel.api.InternalIndexState;
+import org.neo4j.internal.kernel.api.SchemaRead;
+import org.neo4j.internal.kernel.api.TokenNameLookup;
+import org.neo4j.internal.kernel.api.TokenRead;
 import org.neo4j.internal.kernel.api.exceptions.LabelNotFoundKernelException;
 import org.neo4j.internal.kernel.api.exceptions.schema.IndexNotFoundKernelException;
 import org.neo4j.internal.kernel.api.schema.constraints.ConstraintDescriptor;
 import org.neo4j.kernel.api.KernelTransaction;
 import org.neo4j.kernel.api.SilentTokenNameLookup;
 import org.neo4j.kernel.api.Statement;
-import org.neo4j.procedure.*;
+import org.neo4j.procedure.Context;
+import org.neo4j.procedure.Description;
+import org.neo4j.procedure.Mode;
+import org.neo4j.procedure.Name;
+import org.neo4j.procedure.Procedure;
+import org.neo4j.procedure.UserFunction;
 
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
 import java.util.concurrent.ExecutionException;
 import java.util.stream.Collectors;
 import java.util.stream.Stream;
@@ -417,6 +434,12 @@ public class Schemas {
         List<String> properties = new ArrayList<>();
         Arrays.stream(indexReference.properties()).forEach((i) -> properties.add(tokens.propertyKeyGetName(i)));
         try {
+            float populationProgressValue = db.execute("CALL db.indexes() YIELD tokenNames, properties, progress " +
+                    "WHERE $label IN tokenNames AND $properties = properties " +
+                    "RETURN progress", Util.map("label", labelName, "properties", properties))
+                    .<Double>columnAs("progress")
+                    .next()
+                    .floatValue();
             return new IndexConstraintNodeInfo(
                     // Pretty print for index name
                     String.format(":%s(%s)", labelName, StringUtils.join(properties, ",")),
@@ -425,7 +448,7 @@ public class Schemas {
                     schemaRead.indexGetState(indexReference).toString(),
                     !indexReference.isUnique() ? "INDEX" : "UNIQUENESS",
                     schemaRead.indexGetState(indexReference).equals(InternalIndexState.FAILED) ? schemaRead.indexGetFailure(indexReference) : "NO FAILURE",
-                    schemaRead.indexGetPopulationProgress(indexReference).getCompleted() / schemaRead.indexGetPopulationProgress(indexReference).getTotal() * 100,
+                    populationProgressValue,
                     schemaRead.indexSize(indexReference),
                     schemaRead.indexUniqueValuesSelectivity(indexReference),
                     indexReference.userDescription(tokens)

